systemd-swap

```diff
@@ -190,41 +194,55 @@ case "$1" in
-                                                        ERRO "Zram: this kernel does not support hot add zram device, please use 4.2+ kernels or see modinfo zram and make a modprobe rule"
-                                                read -r NEW_ZRAM < /sys/class/zram-control/hot_add
-                                                INFO "Zram: success: new device /dev/zram$NEW_ZRAM"
-                                        ;;
-                                        /dev/zram*)
-                                                [ -b "$OUTPUT" ] || continue
-                                                zram_dev="$OUTPUT"
-                                                break
-                                        ;;
-                                esac
-                        done
-                        if [ -b "$zram_dev" ]; then
-                            INFO "Zram: initialized: $zram_dev"
-                            mkswap "$zram_dev" &> /dev/null && \
-                            UNIT_NAME=$(gen_swap_unit What="$zram_dev" Options=discard Priority=$zram_prio Tag=zram)
-                            systemctl daemon-reload
-                            systemctl start $UNIT_NAME
-                        else
-                            WARN "Can't get free Zram device"
-                        fi
+gen_zram_device(){
+
+  INFO "Zram: trying to initialize free device"
+  for (( i = 0; i < 10; i++ )); do
+          # zramctl is a external program -> return name of first free device
+          TMP=$(mktemp)
+          zramctl -f -a $zram_alg -t $zram_wstreams -s $zram_size &> $TMP
+          read -r OUTPUT < $TMP
+          rm -f $TMP
+          case "$OUTPUT" in
+                  *"failed to reset: Device or resource busy"*)
+                          sleep 1
+                  ;;
+                  *"zramctl: no free zram device found"*)
+                          WARN "Zram: zramctl can't find free device"
+                          INFO "Zram: using workaround hook for hot add"
+                          [ ! -f /sys/class/zram-control/hot_add ] && \
+                                  ERRO "Zram: this kernel does not support hot add zram device, please use 4.2+ kernels or see modinfo zram and make a modprobe rule"
+                          read -r NEW_ZRAM < /sys/class/zram-control/hot_add
+                          INFO "Zram: success: new device /dev/zram$NEW_ZRAM"
+                  ;;
+                  /dev/zram*)
+                          [ -b "$OUTPUT" ] || continue
+                          zram_dev="$OUTPUT"
+                          break
+                  ;;
+          esac
+  done
+
+  if [ -b "$zram_dev" ]; then
+    return;
+  else
+    ERRO "Zram: failed to init device"
+  fi
+}
+
+                        for ((dev = 0; dev < zram_rstreams; dev++)); do
+                              gen_zram_device
+
+                              if [ -b "$zram_dev" ]; then
```

Comment:
Okay, if you wish, you can move logic for get_zram_dev in function, but:

- No inplace definition of such big function, or try honor indentation
- Function return something, if you wish use of vars defined in function.
  that's not obvious what you define zram_dev in function and what zram_bdev defined somewhere as global.

Reply:
Got it, agreed